openapi.yaml

```diff
@@ -0,0 +1,92 @@
+openapi: 3.1.0
+info:
+  version: 0.1.0
+  title: AsyncAPI Server API
+  description: Server API providing official AsyncAPI tools
+  contact:
+    name: AsyncAPI Initiative
+    email: info@asyncapi.io
+    url: https://asyncapi.com/
+  license:
+    name: Apache 2.0
+    url: https://www.apache.org/licenses/LICENSE-2.0.html
+
+servers:
+  - url: http://api.asyncapi.io
+
+paths:
+  /generate:
```

Comment:
Hmm. This idea with `generate/template` is good, then we can extend it by `generate/models` and other things to split functionality to separate paths.

About versioning, we can go with `alpha` in first version and then switch to v1, or go with `v1` from beginning. To be honest I was writing backend servers just only for fun and don't know what the best practices are.

Reply:
Not sure if makes sense to use different servers for this? Like one server with the suffix `/v1`, etc. 
But anyway, I think it is ok like you did now.

Reply:
@boyney123

I was also thinking about adding a parameter to the path, but I don't know if it's such a good option when we'll be serving all our templates. As for custom templates, for security reasons, we will have to add them manually to the supported templates - so the user will have a finite number of templates to generate anyway. Let's see what others will say. Maybe the template name as parameter is a good idea and I'm wrong 😅

Reply:
Maybe `/generate/template/{template}` for official ones, `/generate/template/custom?template={template_url}` or similar?

Reply:
yo, some serious party pooper here 😄 

- isn't `generate` a verb?
- I think templateId should be in the path really

I'm seriously concerned about the generation speed and making it synchronous request, where you will end up keeping a connection with the server open for long. Anyway, we are in good position here to not be so worried about first API design. I would recommend try writing the server with https://github.com/anttiviljami/openapi-backend as it is Glee counterpart in OpenAPI world and if we use it actively, we might want to port best features to Glee. Just an idea, to try something different.

Last but not least, I recommend not doing such PR with all the GH workflows as in the end you can go to https://github.com/asyncapi/.github/actions/workflows/global-workflows-support.yml and trigger workflow and it will create PR with all new workflows in repos that do not have it yet, like such a new repo 😉 (yeah, I know, kinda hidden knowledge, yet another one)

Reply:
@BOLT04 Hi! Thanks for this awesome comment! I really appreciate that! I know that there are 3 ways to make versioning:
- URI Versioning
- Header Versioning
- Media Type Versioning - by `Accept` header

As you described the problem with hypermedia and our API users - it may turn out that the effort is not proportional to the benefit. It also seems to me that versioning via URIs (i.e. `v1`, `v2` etc) is usually used rather than headers. I don't know about others opinions, but I'd rather leave the URI versioning in place first, and then think about media type versioning. The first iteration of ServerAPI is just one simple path with templates generation. What do you guys think about it? @smoya @BOLT04 @boyney123 @derberg

Reply:
> The first iteration of ServerAPI is just one simple path with templates generation. What do you guys think about it?

I agree, let's use URI versioning and simplify 👍 

> I think templateId should be in the path really

For me `/generate/template/html` and `/generate/template/markdown` are two different resources in the API, because a URI identifies the resource, therefore those are **two identifiers**. In this case, I think there is only one resource (for now) which is the "Generator". This resource supports one operation, a POST request, where the type of template it should generate is sent in the HTTP body or query parameter. Since it's just a parameter I don't see a problem in using the query parameters for this.

I agree with @derberg and maybe we shouldn't name it `generate`, since it's a verb and usually those aren't meant to be in the path of the URI. The HTTP method is the verb, `POST` is the same as saying "server, please **process** this data representation I'm sending". Maybe it would make sense to be `/generator/template`, but then would we have more options for the client to generate? That URI could tell me there is a resource "generator" and a sub-resource inside it called "template". Now, will there be more sub-resources?
From what I know, the AsyncAPI generator calls a template anything that can be generated, so the Generator only generates templates, or does it do more?

With that said, I agree with adding more information and keeping it `generate/template` 👍 . I'm a bit reluctant with the templateId as a path parameter, I'm not against it but I don't see a big justification to do so. If it's for documentation purposes, we can document available template values in the body payload 🙂. But again, I'm comfortable going forward because this PR already provides a lot of value 💯 !

Reply:
I have the feeling that REST is not the best architectural style for this API. Have you considered going with RPC style? Have a look: https://www.smashingmagazine.com/2016/09/understanding-rest-and-rpc-for-http-apis/

We can have endpoints like:

```
POST /v1/generateHtml
POST /v1/generateMarkdown
POST /v1/generateModel
```

It's pretty much like a CLI but remote.

Reply:
Then just use another model like `/generate/:templateName`, `/convert/220/to/300` or similar. Read a bit about RPC over HTTP. Not everything has to be REST or GraphQL 😉

Reply:
`/generate/:templateName` was discussed in this thread and we decide to go with `/generator` and template will as one of the payload's fields.

For `/convert/220/to/300` I think that having `from` and `to` versions in request body will be better 😄

Reply:
Ok, just a thought: if you go for `/convert` with `from` and `to` as payload fields, you should go for `/generate` and not `/generator`.

Reply:
Damn, missed that comment 😅  Yeah, right, but instead of describing resources as a verb we should describe as a noun (we discuss about it in this thread), so it would be `/converter` rather than `/convert`. Thanks for the observation!

Reply:
Yeah, I saw the convo. I just don't think converter and generator are resources. What is POST /generator supposed to do? Create a generator? No, right? I think your design is more RPC than REST, which is fine, just don't stick to REST rules if it's not REST. My two cents.